### DISCUSSION TITLE
Using multiple cores with OpenMP with M1 Macs

### DISCUSSION THREAD

**Discussion by john-harrold** [nlmixr2/nlmixr2#26] [Show and tell]
The binaries for the M1 Mac on CRAN seem to be built in an environment where OpenMP is not properly configured. The problem is that if you install OpenMP on your Mac correctly, you still won't be able to use it if you use the pre compiled packages from CRAN. After you've configured OpenMP, it's necessary to install `rxode2` from source. This problem will also extend to the `data.table` package as well. 

So the first thing you need to do is configure OpenMP from the instructions
here: https://mac.r-project.org/openmp/

You can test this using the following code taken from [stack overflow](https://stackoverflow.com/questions/70638118/configuring-compilers-on-mac-m1-big-sur-monterey-for-rcpp-and-other-tools):

```r
if (!requireNamespace("RcppArmadillo", quietly = TRUE)) {
    install.packages("RcppArmadillo")
}
Rcpp::sourceCpp(code = '
#include <RcppArmadillo.h>
#ifdef _OPENMP
# include <omp.h>
#endif

// [[Rcpp::depends(RcppArmadillo)]]
// [[Rcpp::export]]
void omp_test()
{
#ifdef _OPENMP
    Rprintf("OpenMP threads available: %d\\n", omp_get_max_threads());
#else
    Rprintf("OpenMP not supported\\n");
#endif
}
')
omp_test()
```

When you run that it should give you something like:

```r
OpenMP threads available: 10
```

If that works you should then install `rxode2` and `data.table` from source:

```r
install.packages("data.table", type="source")
install.packages("rxode2", type="source")
install.packages("nlmixr2est", type="source")
```

**Comment by mattfidler** [comment]
Also to be safe, install `nlmixr2est` from source to make sure that the binaries are compatible:

```r
install.packages("nlmixr2est", type="source")
```

**Reply by john-harrold** [reply]
I added it above.

**Comment by mattfidler** [comment]
Great thanks!

**Comment by Chickenlover0908** [comment]
Thank you very much!!!🌹🌹🌹🌹
It have spend me many time to set openmp.
Only your approach is useful.
I almost gave up.

**Comment by sliao999** [comment]
Would this work for Windows based desktop?

**Reply by mattfidler** [reply]
Openmp should work for windows by simply using the cran binaries
